src/main/scala/scalaz/stream/io.scala

```diff
@@ -212,4 +217,173 @@ object io {
         else buf.take(m)
       }}
     }
+
+  /**
+   * Converts a source to a mutable `InputStream`.  The resulting input stream
+   * should be reasonably efficient and supports early termination (i.e. all
+   * finalizers associated with the input process will be run if the stream is
+   * closed).
+   *
+   * Takes a `copy` function which itself takes the following parameters in order:
+   *
+   * - `A` - the source chunk
+   * - `Int` - the logical offset into the source chunk
+   * - `Array[Byte]` - the target copy buffer
+   * - `Int` - the logical offset into the target buffer
+   * - `Int` - the length to copy
+   *
+   * The `copy` function returns the number of bytes copied, in total.  You should
+   * think of this function as a generic version of `System.arraycopy`, which is
+   * in fact the precise implementation you would use if `A = Array[Byte]`.  Note
+   * that `toInputStream` will never request a `copy` with a length value that
+   * exceeds the logical length of the source chunk as reported by the `logicalLength`
+   * function.  Additionally, `buffer` will always be *at least* `off2 + length` in
+   * size.  Thus, the `copy` function does not need to concern itself with bounds
+   * checking.
+   *
+   * The other function taken as a parameter is `logicalLength`, which simply returns
+   * the logical length of the source chunk in question.  Thus, if `A = Array[Byte]`,
+   * this function would simply be `{ _.length }`.
+   *
+   * @see #toInputStreamFromBytes
+   */
+  def toInputStream[A](p: Process[Task, A])(logicalLength: A => Int, copy: (A, Int, Array[Byte], Int, Int) => Int): InputStream = new InputStream {
```

Comment:
Random thought: should this be a typeclass instead?  Among other things, a typeclass would be more PIC-friendly, and so it would actually be substantially faster than random functions.  It would also be quite a bit cleaner.  The obvious downside being that it would be more opaque, and also a typeclass of somewhat dubious soundness.